apis/cloudflare/src/realtime-logger.ts

```diff
@@ -0,0 +1,772 @@
+        i++;
+      }
+      this.audioBuffers = this.audioBuffers.slice(i + 1);
+      this.baseTimeMs +=
+        (trimmedBytes / (this.bytesPerSample * this.inputCodec.sample_rate)) *
+        1000;
+      console.warn(
+        `Trimmed ${trimmedBytes} bytes from audio buffer; now ${this.totalByteLength} bytes total`,
+      );
+    }
+  }
+
+  trimStart(startTimeMs: number) {
+    let startByteIndex = this.timestampToByteIndex(startTimeMs);
+
+    let i = 0;
+    while (
+      i < this.audioBuffers.length &&
+      startByteIndex > this.audioBuffers[i].byteLength
+    ) {
+      startByteIndex -= this.audioBuffers[i].byteLength;
+      i++;
+    }
+    this.audioBuffers = this.audioBuffers.slice(i + 1);
+
+    if (this.audioBuffers.length) {
+      this.audioBuffers[0] = this.audioBuffers[0]?.slice(startByteIndex);
+    }
+
+    this.baseTimeMs = startTimeMs;
+  }
+
+  get byteLength(): number {
+    return this.totalByteLength;
+  }
+
+  encode(compress: boolean, endTimeMs?: number): [Blob, string] {
+    let slicedBuffers = this.audioBuffers;
+    let totalByteLength = this.totalByteLength;
+    if (endTimeMs) {
+      let endByteIndex = this.timestampToByteIndex(endTimeMs);
+      let i = this.audioBuffers.length - 1;
+      while (i >= 0 && endByteIndex > totalByteLength) {
+        totalByteLength -= this.audioBuffers[i].byteLength;
+        i--;
+      }
+      slicedBuffers = this.audioBuffers.slice(0, i + 1);
+      if (slicedBuffers.length) {
+        const end = slicedBuffers.length - 1;
+        slicedBuffers[end] = slicedBuffers[end].slice(
+          0,
+          totalByteLength - endByteIndex + 1,
+        );
+      }
+    }
+
+    if (compress && this.inputCodec.name !== "g711") {
+      return [makeMp3File(this.inputCodec, 40, slicedBuffers), "mp3"];
+    } else {
+      return [makeWavFile(this.inputCodec, slicedBuffers), "wav"];
+    }
+  }
+
+  private timestampToByteIndex(timestampMs: number): number {
+    return (
+      Math.round(
+        ((timestampMs - this.baseTimeMs) / 1000) * this.inputCodec.sample_rate,
+      ) * this.bytesPerSample
+    );
+  }
+}
+
+function openAiToPcmAudioFormat(audioFormat: AudioFormatType): PcmAudioFormat {
+  const common = {
+    byte_order: "little",
+    number_encoding: "int",
+    channels: 1,
+  } as const;
+  // https://platform.openai.com/docs/guides/realtime/audio-formats
+  switch (audioFormat) {
+    case "pcm16":
+      return {
+        ...common,
+        name: "pcm",
+        bits_per_sample: 16,
+        sample_rate: 24000,
+      };
+    case "g711_ulaw":
+      return {
+        ...common,
+        name: "g711",
+        algorithm: "mu",
+        sample_rate: 8000,
+      };
+    case "g711_alaw":
+      return {
+        ...common,
+        name: "g711",
+        algorithm: "a",
+        sample_rate: 8000,
+      };
+    default:
+      const x: never = audioFormat;
+      throw new Error(`Unknown audio format ${JSON.stringify(x)}`);
+  }
+}
+
+/**
```

Comment:
i wonder if we shuold move this into the SDK (perhaps in a separate change). So you can use it outside of the proxy too 🤔

Reply:
Will do in a separate change. Probably also needs some API cleanup before we can commit to maintaining it as part of the SDK.